mlflow/langchain/_langchain_autolog.py

```diff
@@ -85,26 +86,50 @@ def _invoke(self, *args, **kwargs):
             return original(self, *args, **kwargs)
 
     config = AutoLoggingConfig.init(mlflow.langchain.FLAVOR_NAME)
-
-    if not IS_PATCHING_DISABLED_FOR_TRACING.get() and config.log_traces:
-        args, kwargs = _get_args_with_mlflow_tracer(func_name, args, kwargs)
+    should_trace = not IS_PATCHING_DISABLED_FOR_TRACING.get() and config.log_traces
+
+    if should_trace:
+        tracer = MlflowLangchainTracer(
+            # NB: RunnableSequence's batch() and abatch() methods are implemented in a peculiar way
+            # that iterates on steps->items sequentially within the same thread. For example, if a
+            # sequence has 2 steps and the batch size is 3, the execution flow will be:
+            #  - Step 1 for item 1
+            #  - Step 1 for item 2
+            #  - Step 1 for item 3
+            #  - Step 2 for item 1
+            #  - Step 2 for item 2
+            #  - Step 2 for item 3
+            # Due to this behavior, we cannot attach the span to the context for this particular
+            # API, otherwise spans for different inputs will be mixed up.
+            set_span_in_context=not (
+                isinstance(self, RunnableSequence) and func_name in ["batch", "abatch"]
+            )
+        )
+        args, kwargs = _get_args_with_mlflow_tracer(tracer, func_name, args, kwargs)
 
     # Traces does not require an MLflow run, only the other optional artifacts require it.
-    if not IS_PATCHING_DISABLED_FOR_ARTIFACTS and config.should_log_optional_artifacts():
-        with _setup_autolog_run(config, self) as run_id:
+    try:
+        if not IS_PATCHING_DISABLED_FOR_ARTIFACTS and config.should_log_optional_artifacts():
+            with _setup_autolog_run(config, self) as run_id:
+                result = _invoke(self, *args, **kwargs)
+                _log_optional_artifacts(config, run_id, result, self, func_name, *args, **kwargs)
+        else:
             result = _invoke(self, *args, **kwargs)
-            _log_optional_artifacts(config, run_id, result, self, func_name, *args, **kwargs)
-    else:
-        result = _invoke(self, *args, **kwargs)
+    finally:
+        if should_trace:
```

Comment:
nit: we may not need to check `should_trace` if `flush` does not have side effects.

Reply:
This gating is because we only instantiate the tracer when `should_trace` is `True`